src/util/data.rs

```diff
@@ -0,0 +1,62 @@
+use std::fmt;
+use std::str;
+
+use serde::Serialize;
+use snafu::Snafu;
+
+#[derive(Debug)]
+pub enum ProjectId {
+    NamespaceSlug { namespace: String, slug: String },
+    Id(String),
+    Url(String),
+}
+
+impl ProjectId {
+    pub fn as_string(&self) -> String {
+        match self {
+            ProjectId::NamespaceSlug { namespace, slug } => {
+                format!("{}/{}", namespace, slug)
+            }
+            ProjectId::Id(id) => id.to_string(),
+            ProjectId::Url(url) => url.to_string(),
+        }
+    }
+}
+
+#[derive(Debug, PartialEq, Snafu)]
+pub struct ProjectIdParseError;
+
+impl str::FromStr for ProjectId {
+    type Err = ProjectIdParseError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        if s.starts_with("http") {
+            Ok(ProjectId::Url(s.to_string()))
+        } else {
+            match s.split_once('/') {
+                Some((pre, suf)) => Ok(ProjectId::NamespaceSlug {
+                    namespace: pre.into(),
+                    slug: suf.into(),
+                }),
+                None => Ok(ProjectId::Id(s.to_string())),
+            }
+        }
+    }
+}
+
+impl fmt::Display for ProjectId {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.as_string())
+    }
+}
```

Comment:
suggestion: I think if you implement `Into<String>` you can derive `Display`

Reply:
This didn't work for me. I thought we "should" implement `Display` in general. Is there maybe some crate necessary? The docs about display say

> "Implementing this trait for a type will automatically implement the [ToString](https://doc.rust-lang.org/std/string/trait.ToString.html) trait for the type, allowing the usage of the [.to_string()](https://doc.rust-lang.org/std/string/trait.ToString.html#tymethod.to_string) method. Prefer implementing the Display trait for a type, rather than [ToString](https://doc.rust-lang.org/std/string/trait.ToString.html)."

I might not understand … :-) 
But I can then remove my previously created non-idiomatic `as_string()` of course….

Reply:
I think I might have misremembered how Display works.